packages/utils/src/ast-utils/helpers.ts

```diff
@@ -1,38 +1,40 @@
 import { AST_NODE_TYPES, AST_TOKEN_TYPES, TSESTree } from '../ts-estree';
 
-type ObjectEntry<BaseType> = [keyof BaseType, BaseType[keyof BaseType]];
+type ObjectEntry<BaseType> = BaseType extends unknown
+  ? [keyof BaseType, BaseType[keyof BaseType]]
+  : never;
 type ObjectEntries<BaseType> = Array<ObjectEntry<BaseType>>;
 
 export const isNodeOfType =
   <NodeType extends AST_NODE_TYPES>(nodeType: NodeType) =>
   (
     node: TSESTree.Node | null | undefined,
-  ): node is TSESTree.Node & { type: NodeType } =>
+  ): node is Extract<TSESTree.Node, { type: NodeType }> =>
     node?.type === nodeType;
 
 export const isNodeOfTypes =
   <NodeTypes extends readonly AST_NODE_TYPES[]>(nodeTypes: NodeTypes) =>
   (
     node: TSESTree.Node | null | undefined,
-  ): node is TSESTree.Node & { type: NodeTypes[number] } =>
+  ): node is Extract<TSESTree.Node, { type: NodeTypes[number] }> =>
     !!node && nodeTypes.includes(node.type);
 
 export const isNodeOfTypeWithConditions = <
   NodeType extends AST_NODE_TYPES,
-  Conditions extends Partial<TSESTree.Node & { type: NodeType }>,
+  Conditions extends Partial<Extract<TSESTree.Node, { type: NodeType }>>,
 >(
   nodeType: NodeType,
   conditions: Conditions,
 ): ((
   node: TSESTree.Node | null | undefined,
-) => node is TSESTree.Node & { type: NodeType } & Conditions) => {
+) => node is Extract<TSESTree.Node, { type: NodeType }> & Conditions) => {
   const entries = Object.entries(conditions) as ObjectEntries<TSESTree.Node>;
 
   return (
     node: TSESTree.Node | null | undefined,
-  ): node is TSESTree.Node & { type: NodeType } & Conditions =>
+  ): node is Extract<TSESTree.Node, { type: NodeType }> & Conditions =>
     node?.type === nodeType &&
-    entries.every(([key, value]) => node[key] === value);
+    entries.every(([key, value]) => key in node && node[key] === value);
```

Comment:
I could reproduce it in Playground here – https://tsplay.dev/NByybw

As @Josh-Cena mentioned https://github.com/typescript-eslint/typescript-eslint/pull/4545#discussion_r895570784, `keyof Node` can help here.

It's working after your changes – https://tsplay.dev/mbBBbm

Reply:
Should be fixed now 👍